packages/fast-usdc/src/exos/status-manager.js

```diff
@@ -1,12 +1,205 @@
+import { makeError, q } from '@endo/errors';
+
+import { CctpTxEvidenceShape, StatusManagerValueShape } from '../typeGuards.js';
+import { TxStatus } from '../constants.js';
+
 /**
+ * @import {MapStore} from '@agoric/store';
  * @import {Zone} from '@agoric/zone';
+ * @import {CctpTxEvidence, NobleAddress, StatusManagerKey, StatusManagerValue} from '../types.js';
+ */
+
+/**
+ * @param {NobleAddress} addr
+ * @param {bigint} amount
  */
+export const toStatusKey = (addr, amount) =>
+  /** @type {StatusManagerKey} */ (JSON.stringify([addr, String(amount)]));
+
+/**
+ * @param {CctpTxEvidence} evidence
+ * @returns {StatusManagerKey}
+ */
+export const getStatusKey = evidence => {
+  const { amount, forwardingAddress } = evidence.tx;
+  return toStatusKey(forwardingAddress, amount);
+};
 
 /**
  * @param {Zone} zone
  */
 export const prepareStatusManager = zone => {
-  return zone.exo('Fast USDC Status Manager', undefined, {});
+  /** @type {MapStore<StatusManagerKey, StatusManagerValue[]>} */
+  const txs = zone.mapStore('txs', {
+    keyShape: M.string(),
+    valueShape: M.arrayOf(StatusManagerValueShape),
+  });
+
+  /**  @param {StatusManagerValue[]} entries */
+  const getPendingIdx = entries =>
+    entries.findIndex(
+      ({ status }) =>
+        // TODO since we remove SETTLED, maybe this is not needed
+        // or could become `status !== TxStatus.Observed`
+        status === TxStatus.Advanced || status === TxStatus.Skipped,
+    );
+
+  return zone.exo(
+    'Fast USDC Status Manager',
+    M.interface('StatusManagerI', {
+      observe: M.call(CctpTxEvidenceShape).returns(M.number()),
+      advance: M.call(M.string(), M.bigint(), M.number()).returns(
+        M.undefined(),
+      ),
+      skip: M.call(M.string(), M.bigint(), M.number()).returns(M.undefined()),
+      hasPendingSettlement: M.call(M.string(), M.bigint()).returns(M.boolean()),
+      settle: M.call(M.string(), M.bigint()).returns(M.undefined()),
+      view: M.call(M.string(), M.bigint())
+        .optional(M.number())
+        .returns(
+          M.or(
+            StatusManagerValueShape,
+            M.arrayOf(StatusManagerValueShape),
+            M.undefined(),
+          ),
+        ),
+    }),
+    {
+      /**
+       * Queue a transaction before initiating a transfer. Receive
+       * the entry index in return.
+       *
+       * @param {CctpTxEvidence} evidence
+       * @returns {number}
+       */
+      observe(evidence) {
+        // XXX like `appendToStoredArray`, but returns index. consider
+        // upstreaming the change if it doesn't affect existing consumers
+        const key = getStatusKey(evidence);
+        const entry = { ...evidence, status: TxStatus.Observed };
+        if (!txs.has(key)) {
+          // XXX append to store helper
+          txs.init(key, harden([entry]));
+          return 0;
+        }
+        const current = txs.get(key);
+        // XXX do we need defensive logic to ensure multiple entries don't have
```

Comment:
a fine invariant to assert. it's recoverable by ignoring duplicates. we don't put that burden on the feed because it would have to keep a record of all transactions. this storage can do the job

Reply:
Ok I'll save that for a later PR, but we can use `txHash` + `chainId` to assert uniqueness.

Thinking we'd want a new SetStore to keep track of these, since the txs `MapStore` ejects entries once they're Settled?

Reply:
Implemented in [6883b74](https://github.com/Agoric/agoric-sdk/pull/10406/commits/6883b740f505a674556acfca54422c26b08ad4cb)